subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/MutationValidator.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+/**
+ * Used to validate mutation of an object and its sub-parts.
+ */
+public interface MutationValidator {
+    /**
+     * Check if mutation is allowed.
+     *
+     * @param lenient <code>false</code> if mutation should be completely forbidden,
+     * or <code>true</code> if mutation is allowed, but a deprecation warning should be shown.
+     */
+    void validateMutation(boolean lenient);
```

Comment:
Lenient is perhaps not the best name here. We are using this for two things:

* to signal whether or not a modification is already problematic on an "observed" configuration (`true`), or it should only fail for a "resolved" configuration (`false`)
* to tell if we should throw an error (`false`) or just show a deprecation warning (`true`) if we find a problem

That the two things are covered by the same values is just coincidence. Considering that the second use-case is going away in Gradle 3.0, I would target the name towards the first use-case. Yet, I can't really find a good name for that either.